neuralmonkey/decoders/decoder.py

```diff
@@ -266,42 +239,124 @@ def _create_initial_state(self) -> None:
+        if self.embeddings_source is not None:
+            return self.embeddings_source.embedding_matrix
+
+        # TODO better initialization
+        return tf.get_variable(
+            "word_embeddings", [len(self.vocabulary), self.embedding_size],
+            initializer=tf.random_uniform_initializer(-0.5, 0.5))
+
+    @tensor
+    def decoding_w(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "state_to_word_W", [self.rnn_size, len(self.vocabulary)],
                 initializer=tf.random_uniform_initializer(-0.5, 0.5))
-        else:
-            self.embedding_matrix = self.embeddings_source.embedding_matrix
-
-    def embed_and_dropout(self, inputs: tf.Tensor) -> tf.Tensor:
-        """Embed the input using the embedding matrix and apply dropout
 
-        Arguments:
-            inputs: The Tensor to be embedded and dropped out.
-        """
-        with tf.variable_scope("embed_inputs"):
-            embedded = tf.nn.embedding_lookup(
-                self.embedding_matrix, inputs)
-            return dropout(embedded,
-                           self.dropout_keep_prob,
-                           self.train_mode)
+    @tensor
+    def decoding_b(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "state_to_word_b", [len(self.vocabulary)],
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        # POSLEDNI TRAIN INPUT SE V DEKODOVACI FUNKCI NEPOUZIJE
+        # (jen jako target)
+        logits, _, _ = self._decoding_loop(train_mode=True)
+
+        return logits
+
+    @tensor
+    def runtime_loop_result(self) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        return self._decoding_loop(train_mode=False)
+
+    @tensor
+    def runtime_logits(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        return self.runtime_loop_result[0]
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def runtime_rnn_states(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        return self.runtime_loop_result[1]
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def runtime_mask(self) -> tf.Tensor:
+        # pylint: disable=unsubscriptable-object
+        return self.runtime_loop_result[2]
+        # pylint: enable=unsubscriptable-object
+
+    @tensor
+    def train_xents(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+
+        return tf.contrib.seq2seq.sequence_loss(
+            tf.transpose(self.train_logits, perm=[1, 0, 2]),
+            train_targets,
+            tf.transpose(self.train_padding),
+            average_across_batch=False)
+
+    @tensor
+    def train_loss(self) -> tf.Tensor:
+        return tf.reduce_mean(self.train_xents)
+
```

Comment:
`train_xents` se počítaj na transponovaných `train_inputs`, protože `train_inputs` maj v první dimenzi čas a v druhý batch. `tf.contrib.seq2seq.sequence_loss` žere jako targety tensor, co má v první dimenzi batch a ve druhý čas. Tahle funkce je schopná vracet loss pro každý slovo. My chcem, aby `train_xents` byly lossy pro každou větu. Třeba kvůli nějakejm reinforce hárdkórech. A celej `train_loss` už je pak jen  skalár. Ty `train_xents` jsou umělej mezikrok. Kdyby tam nebylo to `average_across_batch=False`, tak by to bylo to samý jako tadyten loss.